src/ramanchada2/protocols/twinning.py

```diff
@@ -0,0 +1,283 @@
+
+        peak (float): The position of the peak (in nm) of interest for analysis, with a default value of 144 (TiO2).
+
+    Methods:
+        __init__(self, twinned: SpectraFrame, reference: SpectraFrame, boundaries=None, peak_at=144):
+            Initializes a new TwinningComponent object by averaging the provided twinned and reference spectra
+            based on predefined grouping columns. Optionally, boundaries for analysis and a peak position can be
+            specified.
+
+    Parameters:
+        twinned (SpectraFrame): The SpectraFrame representing the twinned data.
+        reference (SpectraFrame): The SpectraFrame representing the reference data.
+        boundaries (tuple, optional): Optional boundary values to restrict the analysis region (default: (50, 2000)).
+        peak_at (int, optional): The peak position to focus the analysis on (default: 144 for TiO2).
+    """
+
+    def __init__(
+        self,
+        twinned: SpectraFrame,
+        reference: SpectraFrame,
+        boundaries=None,
+        reference_band_nm=144.0,
+        grouping_cols=[
+            "sample",
+            "provider",
+            "device",
+            "laser_wl",
+            "laser_power_percent",
+            "laser_power_mW",
+            "time_ms",
+        ],
+    ):
+        self.grouping_cols = grouping_cols
+        self.twinned = twinned.average(grouping_cols=self.grouping_cols)
+        self.reference = reference.average(grouping_cols=self.grouping_cols)
+        self.boundaries = (50, 2000) if boundaries is None else boundaries
+        self.linreg_reference = (None, None)
+        self.linreg_twinned = (None, None)
+        self.correction_factor: float = 1.0
+        self.reference_band_nm = reference_band_nm
+
+    def normalize_by_laserpower_time(self, source="spectrum", target="spectrum"):
+        for (index_refere4nce, row_reference), (index_twinned, row_twinned) in zip(
+            self.reference.iterrows(), self.twinned.iterrows()
+        ):
+            laser_power_ratio = (
+                row_reference["laser_power_mW"] / row_twinned["laser_power_mW"]
+            )
+            time_ratio = row_reference["time_ms"] / row_twinned["time_ms"]
+            spe = row_twinned[source]
+            self.twinned.at[index_twinned, target] = (
+                spe * laser_power_ratio * time_ratio
+            )
+            self.twinned.at[index_twinned, "laser_power_ratio"] = laser_power_ratio
+            self.twinned.at[index_twinned, "time_ratio"] = time_ratio
+
+    def calc_peak_intensity(
+        self,
+        spe: Spectrum,
+        boundaries=None,
+        prominence_coeff=0.01,
+        no_fit=False,
+        peak_intensity="height",
+    ):
+        try:
+            if boundaries is None:
+                boundaries = (self.reference_band_nm - 50, self.reference_band_nm + 50)
+            spe = spe.trim_axes(method="x-axis", boundaries=boundaries)
```

Comment:
mypy: error: "Spectrum" has no attribute "trim_axes"  [attr-defined]

Reply:
this is decorator issue